[Memory Snapshot] Fix RecordFunction Callback Handling

sraikund16 opened the pull request:
Fixes #153571 
Summary:
1. Set annotation callback to global to include all threads
2. Only init callbacks when enable == true and callbacks are empty under mutex
3. When enable == false, check if callbacks are present and if so remove them and set handle to 0 under mutex

We don't expect memory snapshots to be called from several different threads (almost always called just from main) but we make sure to add thread safety in the off case that users do want to call it from different points of entry

Test Plan: Ran basic snapshot and saw that the callbacks were registered properly

Reviewed By: ngimel

Differential Revision: D74771491

Skylion007 commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -15,6 +15,39 @@ using torch::jit::Pickler;
+    // Constructor
+    CallbackManager() : annotationHandle(0), compileContextHandle(0) {}
+    // Destructor
+    ~CallbackManager() {
+        // Release any resources held by the callback handles if necessary
+    }
```
```suggestion
    ~CallbackManager() = default;
```

Skylion007 commented on torch/csrc/cuda/memory_snapshot.cpp:
```diff
@@ -15,6 +15,39 @@ using torch::jit::Pickler;
+class CallbackManager {
+public:
+    // Constructor
+    CallbackManager() : annotationHandle(0), compileContextHandle(0) {}
```
These can be default member initalizer fields

Skylion007 reviewed (approved):
Approved with the ctor default nit

sraikund16 commented:
@eqy @syed-ahmed can I get another review?

sraikund16 committed:
[Memory Snapshot] Fix RecordFunction Callback Handling (#153839)

Summary:

1. Set annotation callback to global to include all threads
2. Only init callbacks when enable == true and callbacks are empty under mutex
3. When enable == false, check if callbacks are present and if so remove them and set handle to 0 under mutex

We don't expect memory snapshots to be called from several different threads (almost always called just from main) but we make sure to add thread safety in the off case that users do want to call it from different points of entry

Test Plan: Ran basic snapshot and saw that the callbacks were registered properly

Reviewed By: ngimel

Differential Revision: D74771491

sraikund16 commented:
@pytorchbot  merge
